packages/focal/src/react/react.ts

```diff
@@ -76,50 +93,45 @@ export class LiftWrapper<TProps>
       // Could this be replaced by a regular closure? Perhaps using
       // a class is an optimization?
       case 1:
-        new RenderOne(this, newProps) // eslint-disable-line
+        new RenderOne(renderCache, newProps, setCache, setSubscription) // eslint-disable-line
         break
       default:
-        new RenderMany(this, newProps, n) // eslint-disable-line
+        new RenderMany(renderCache, newProps, setCache, setSubscription, n) // eslint-disable-line
         break
     }
-  }
 
-  private _unsubscribe() {
-    const subscription = this.state ? this.state.subscription : null
-    if (subscription)
-      subscription.unsubscribe()
+    inSync = true
   }
 
-  // eslint-disable-next-line camelcase
-  UNSAFE_componentWillReceiveProps(newProps: LiftWrapperProps<TProps>) {
-    this._unsubscribe()
-    this._subscribe(newProps)
-  }
+  export const Renderer = <TProps>(props: Props<TProps>) => {
+    const _subscription = React.useRef<WrapperSubscription>(null)
+    const _renderCache = React.useRef<RenderCache>(null)
 
-  // eslint-disable-next-line camelcase
-  UNSAFE_componentWillMount() {
-    this._unsubscribe()
-    this._subscribe(this.props)
-  }
+    // concurrent mode support
+    // see https://codesandbox.io/s/x2p46v02z4?from-embed=&file=/src/BadCounter.jsx
+    const subscription = { current: _subscription.current }
+    const renderCache = { current: _renderCache.current }
 
-  componentWillUnmount() {
-    this._unsubscribe()
-    this.setState(LiftWrapper._initState)
-  }
+    React.useEffect(() => {
+      _subscription.current = subscription.current
+      _renderCache.current = renderCache.current
+    })
 
-  shouldComponentUpdate(
-    _newProps: Readonly<LiftWrapperProps<TProps>>,
-    newState: Readonly<LiftWrapperState>,
-    _newContext: any
-  ) {
-    return newState.renderCache !== this.state.renderCache
+    React.useEffect(() => () => unsubscribe(subscription), [])
+
+    unsubscribe(subscription)
```

Comment:
Not sure, but seems like a redundant `unsubscribe`, it should be handled on "unmount" in `React.useEffect`. Am I right?

Reply:
No relevant anymore